resourcemanager/pool/workerpool/workerpool.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workerpool
+
+import (
+	"time"
+
+	"github.com/pingcap/tidb/resourcemanager"
+	"github.com/pingcap/tidb/resourcemanager/util"
+	tidbutil "github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/syncutil"
+	atomicutil "go.uber.org/atomic"
+)
+
+// Worker is worker interface.
+type Worker[T any] interface {
+	HandleTask(task T)
+	Close()
+}
+
+// WorkerPool is a pool of workers.
+type WorkerPool[T any] struct {
+	name         string
+	numWorkers   int32
+	runningTask  atomicutil.Int32
+	taskChan     chan T
+	quitChan     chan struct{}
+	wg           tidbutil.WaitGroupWrapper
+	createWorker func() Worker[T]
+	lastTuneTs   atomicutil.Time
+	mu           syncutil.RWMutex
+}
+
+// NewWorkerPool creates a new worker pool.
+func NewWorkerPool[T any](name string, component util.Component, numWorkers int, createWorker func() Worker[T]) (*WorkerPool[T], error) {
+	if numWorkers <= 0 {
+		numWorkers = 1
+	}
+
+	p := &WorkerPool[T]{
+		name:         name,
+		numWorkers:   int32(numWorkers),
+		taskChan:     make(chan T),
+		quitChan:     make(chan struct{}),
+		createWorker: createWorker,
+	}
+
+	err := resourcemanager.InstanceResourceManager.Register(p, name, component)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start default count of workers.
+	for i := 0; i < int(p.numWorkers); i++ {
+		p.runAWorker()
+	}
+
+	return p, nil
+}
+
+func (p *WorkerPool[T]) runAWorker() {
+	p.wg.Run(func() {
+		w := p.createWorker()
+		for {
+			select {
+			case task := <-p.taskChan:
+				p.runningTask.Add(1)
+				w.HandleTask(task)
```

Comment:
Should we add ```recovery```here to avoid panic?

Reply:
let the worker to handle it.

Reply:
If wokrer forgets to handle it, how to ensure the correctness here.